### ISSUE TITLE
Cannot use Replication. Error: "cannot unmarshal object into Go value"

### ISSUE THREAD

**Issue by mkboudreau** [jfrog/terraform-provider-artifactory#718]
Artifactory Version: 7.29.9 rev 72909900
License: ProX
Terraform version: 1.4.4
Provider version: 6.35.0

I have tried to do a terraform import with juust about every replication resource. I get the same error on every one.
```
| Error: json: cannot unmarshal object into Go value of type []replication.getLocalSingleReplicationBody
or 
│ Error: json: cannot unmarshal object into Go value of type []replication.getReplicationBody
```
I am guessing this is related to the use of `/api/replications` vs `/api/replications/multiple`. It seems like most of the provider code is using the _multiple_ endpoint in most cases. 

To work around it, I figured, maybe I could try creating the resource using the `artifactory_replication_config` which appears to match our version of Artifactory; however, instead of using the `/api/replications` (which can be used by a ProX license) to create the resource, it is trying to use `/api/replications/multiple` which is for enterprise only. The rest of the endpoints for that resource use `/api/replications` 

```
│ 403 PUT http://xxxx.xxxx.xxxx/artifactory/api/replications/multiple/myrepo
│ {
│   "errors" : [ {
│     "status" : 403,
│     "message" : "Multi-push replication is only available with an Enterprise license."
│   } ]
│ }
│
```  

https://github.com/jfrog/terraform-provider-artifactory/blob/2a85ff6aaba173dc413649127e9b42bc09be88e9/pkg/artifactory/resource/replication/resource_artifactory_local_repository_single_replication.go#L44

https://github.com/jfrog/terraform-provider-artifactory/blob/2a85ff6aaba173dc413649127e9b42bc09be88e9/pkg/artifactory/resource/replication/resource_artifactory_replication_config.go#L242

**Comment by danielmkn** [comment]
Hi @mkboudreau!
`artifactory_replication_config` resource is [deprecated](https://github.com/jfrog/terraform-provider-artifactory/blob/6b4d96a9e066a09d14fb28199ea7ee38c5a23f7d/docs/resources/replication_config.md?plain=1#L6). Also, as you noticed, it's using the endpoint, only available with Ent+ license: `/api/replications/multiple`. That was the reason to add new resources, to make it work with the ProX license as well. The difference is with the Ent+ license you can create [multiple replications](https://jfrog.com/help/r/jfrog-rest-apis/create-or-replace-local-multi-push-replication) and with ProX only [one at a time](https://jfrog.com/help/r/jfrog-rest-apis/update-repository-replication-configuration). 

For your scenario with the ProX license, if you need to import local replication please use `artifactory_local_repository_single_replication`. 
For remote repositories use `artifactory_remote_repository_replication.go`

I was able to import resources using the configuration below. The precondition is these resources should exist in your Artifactory instance, so create the local generic repository with the name `replication-local-generic` and add replication configuration on the Replications tab. 

```
resource "artifactory_local_generic_repository" "replication-local-generic" {
  key = "replication-local-generic"
}

resource "artifactory_local_repository_single_replication" "replication-local-generic" {
  repo_key = "replication-local-generic"
}
```
Run this command in your terminal:
```
$ terraform import artifactory_local_generic_repository.replication-local-generic replication-local-generic

artifactory_local_generic_repository.replication-local-generic: Importing from ID "replication-local-generic"...
artifactory_local_generic_repository.replication-local-generic: Import prepared!
  Prepared artifactory_local_generic_repository for import
artifactory_local_generic_repository.replication-local-generic: Refreshing state... [id=replication-local-generic]

Import successful!

The resources that were imported are shown above. These resources are now in
your Terraform state and will henceforth be managed by Terraform.
```
Then import the replication: 
```
$ terraform import artifactory_local_repository_single_replication.replication-local-generic replication-local-generic

artifactory_local_repository_single_replication.replication-local-generic: Importing from ID "replication-local-generic"...
artifactory_local_repository_single_replication.replication-local-generic: Import prepared!
  Prepared artifactory_local_repository_single_replication for import
artifactory_local_repository_single_replication.replication-local-generic: Refreshing state... [id=replication-local-generic]

Import successful!

The resources that were imported are shown above. These resources are now in
your Terraform state and will henceforth be managed by Terraform.
```

After the import my statefile looks like this: 

```
{
  "version": 4,
  "terraform_version": "1.3.7",
  "serial": 2,
  "lineage": "14687fd2-fd53-6893-b847-a087ed5c2f8e",
  "outputs": {},
  "resources": [
    {
      "mode": "managed",
      "type": "artifactory_local_generic_repository",
      "name": "replication-local-generic",
      "provider": "provider[\"registry.terraform.io/jfrog/artifactory\"]",
      "instances": [
        {
          "schema_version": 0,
          "attributes": {
            "archive_browsing_enabled": false,
            "blacked_out": false,
            "cdn_redirect": false,
            "description": "",
            "download_direct": false,
            "excludes_pattern": "",
            "id": "replication-local-generic",
            "includes_pattern": "**/*",
            "key": "replication-local-generic",
            "notes": "",
            "package_type": "generic",
            "priority_resolution": false,
            "project_environments": [],
            "project_key": "",
            "property_sets": [
              "artifactory"
            ],
            "repo_layout_ref": "simple-default",
            "xray_index": false
          },
          "sensitive_attributes": [],
          "private": "<redacted>"
        }
      ]
    },
    {
      "mode": "managed",
      "type": "artifactory_local_repository_single_replication",
      "name": "replication-local-generic",
      "provider": "provider[\"registry.terraform.io/jfrog/artifactory\"]",
      "instances": [
        {
          "schema_version": 0,
          "attributes": {
            "check_binary_existence_in_filestore": true,
            "cron_exp": "0 0 * * * ?",
            "enable_event_replication": true,
            "enabled": true,
            "exclude_path_prefix_pattern": "/some-other-repo/",
            "id": "replication-local-generic",
            "include_path_prefix_pattern": "/some-repo/",
            "password": null,
            "proxy": "",
            "replication_key": "replication-local-gecd34beeffa",
            "repo_key": "replication-local-generic",
            "socket_timeout_millis": 16000,
            "sync_deletes": true,
            "sync_properties": true,
            "sync_statistics": true,
            "url": "http://localhost.charlesproxy.com:8082",
            "username": "admin"
          },
          "sensitive_attributes": [],
          "private": "<redacted>"
        }
      ]
    }
  ],
  "check_results": null
}
```

**Comment by mkboudreau** [comment]
When I attempted to use `artifactory_local_repository_single_replication`, I got the json unmarshall error. 
```
| Error: json: cannot unmarshal object into Go value of type []replication.getLocalSingleReplicationBody
```

**Comment by mkboudreau** [comment]
I am guessing my version of artifactory is returning the flat response; whereas, the terraform provider is expecting the array.

**Comment by danielmkn** [comment]
Could be, we are testing v6.x on Artifactory 7.49.8 and this new resource for ProX license was added recently, so it was only tested on 7.49.8
I'll test it on your version today and will let you know. We don't have Artifactory 7.29.9 image that works locally on M1, that's why I've tested on 7.49.8

**Comment by danielmkn** [comment]
Hey @mkboudreau, I'm running the same test on 7.29.9 and on 7.24.3 and I'm getting the same response from the [Get replication config](https://jfrog.com/help/r/jfrog-rest-apis/get-repository-replication-configuration) call. Note: the documentation has an incorrect response example (no list), which will be fixed soon. I'm not sure when the change with the list was introduced.

In my 7.29.9/7.24.3 it's a list of one item: 
```
GET /api/replications/generic-local-test
[
    {
        "url": "https://repo1.maven.org/maven2/",
        "socketTimeoutMillis": 15000,
        "username": "admin",
        "password": "",
        "syncStatistics": false,
        "enabled": true,
        "cronExp": "0 0 * * * ?",
        "syncDeletes": false,
        "syncProperties": true,
        "repoKey": "generic-local-test",
        "replicationKey": "generic-local-test_h4a4fe8bdcb",
        "enableEventReplication": false,
        "checkBinaryExistenceInFilestore": false
    }
]
```
A couple of questions: 
1. Can you please give me an example of the response from the same call from your instance?
2. Are you planning to get your Artifactory instance updated any time soon? 

I'm not sure why the response body can be not a list in this case.
Theoretically, if this problem persists and we can reproduce it we might think of a patch for older versions to parse both JSON object and a list of one item for that call.

**Comment by danielmkn** [comment]
It's a good question, I've tried on`Enterprise Plus Trial` and on `ProX` with 7.29.9 and I see the same one-item list.
It's odd. If you want, we can set up a call to see what's wrong, looks like we are in the same timezone.

**Comment by mkboudreau** [comment]
Cool thank you. I am doing some rebuilds right now and may or may not actually need this fixed. I will double-check the json response on the rebuild, both at the 7.29.9 version and again upon upgrading. I'm juggling several things, so it might be a couple days. If I need to set up a call, should I go through my account rep and have him arrange it with you?

**Comment by danielmkn** [comment]
Yes, please, go through the account rep and ask to add me to the call - Partner Engineering, Daniel Miakotkin.

**Comment by mkboudreau** [comment]
Hi @danielmkn, I abandoned using terraform for this and just went to use bash/curl. Hitting the API endpoint with a GET request at `/api/replications/<replication key>`, here is what I discovered:
- New install of Arti 7.29.9 and EntX trial license gave me a json array of objects
- Upgraded to Arti 7.55.10 still with EntX trial license gave me a json array of objects
- Downgraded the license on Arti 7.55.10 to a CommercialX/ProX license and now I get a single json object as a response

**Comment by danielmkn** [comment]
Thanks, @mkboudreau! The Artifactory team told me this recently and also, the [documentation](https://jfrog.com/help/r/jfrog-rest-apis/get-repository-replication-configuration) is fixed now. My problem was that I was given incorrect licenses for testing. In fact, they all were Ent+. It's great you solved your problem, we will check if it's something we should fix on the provider side.

**Comment by oyvsi** [comment]
Also hitting this issue with with Artifactory 7.55.13 with CommercialX license. As @mkboudreau has showed the response from `replications/<key>` is just a single JSON-object. Which I guess makes sense since this license only allows for one URL to be configured. The best way to fix this might be for the API to return a list of one object regardless of the license, but it would be nice if it could be worked around in the provider until then.

**Comment by chb0github** [comment]
@danielmkn I stumbled on a similar situation in another part of the code base (but i don't remember where). What i did was load raw data and marshal it to `interface{}` and then type inspect it. If it was a slice type, remarshalled it the right type as a slice, if not, i marshalled it to a single object of the right and added it to an empty slice. 

Basically I normalized it through type inspection
